design/backup-resources-order.md

```diff
@@ -0,0 +1,39 @@
+## Backup Resources Order
+This document proposes a solution that allows user to specify a backup order for resources of specific resource type.
+
+## Background
+During backup process, user may need to back up resources of specific type in some specific order to ensure the resources were backup properly because these resources are related and ordering might be required to preserve the consistency for the apps to recover itself �from the backup image 
+(Ex: master-slave database pods in a cluster).
+
+## Goals
+- Enable user to specify an order of back up resources belong to specific resource type
+
+## Alternatives Considered
+- Use a plugin to backup an resources and all the sub resources.  For example use a plugin for StatefulSet and backup pods belong to the StatefulSet in specific order.  This plugin solution is not generic and requires plugin for each resource type.
+
+## High-Level Design
+User will specify a map of resource type to list resource names (separate by semicolons).  Each name will be in the format "namespaceName/resourceName" to enable ordering accross namespaces.  Based on this map, the resources of each resource type will be sorted by the order specified in the list of resources.  If a resource instance belong to that specific type but its name is not in the order list, then it will be put behind other resources that are in the list.

```

Comment:
This design relies on the user specifying explicit/exact resource names, is that correct? This is unique within Velero right now, as generally we don't address individual resources. That's not necessarily a blocker, I just wanted to point it out.

What are you envisioning using this with most often? Pods owned by a StatefulSet? I can see the naming being fairly consistent/predictable there, though with pods owned by a Deployment they're less predictable by design, since Kubernetes is supposed to manage them behind the scenes.

Reply:
As of now, we go with the exact match on the Pod names (resources' name).  However, this design may be enhanced to support regular expression search on resources' name.  It may not be as efficient but should be OK.

My usage will only be Pods even though this design support any resource type.  We do a little bit in the preprocessing before sending the sequence of pods down to Velero Backup.  First we fetch the list of Pods, then trace the parents.  Then based on our configuration of the application, we will intelligently generate the sequence the Pods based on the configuration of the application.  For example, if the application comprised of 2 StatefulSets and we need to backup the secondary statefulset first, then we will pick out all pods belong to the secondary StatefulSet and put them on the sequence before putting the pods belong to the primary StatefulSet.  Another example would be if the application comprises of 1 StatefulSet with the pod name "pod-0" is the primary and "pod-1", "pod-2" etc. are secondary then we will generate the sequence "pod-1,pod-2, pod-0" to make sure the secondary pods will be backup before the primary pod.